src/transformers/modeling_gpt2.py

```diff
@@ -693,6 +778,20 @@ def __init__(self, config):
 
         self.init_weights()
 
+        self.model_parallel = False
+
+    def parallelize(self, device_map):
+        self.transformer.parallelize(device_map)
+        self.lm_head = self.lm_head.to(self.transformer.first_device)
+        self.model_parallel = True
```

Comment:
(nit) self.model_parallel is already called in `self.transformer.parallelize(...)` => we could replace it with an `assert self.model_paralell, f"Make sure that model is set in 'parallel' model via `model.model_parallel = True`

Reply:
This one is tricky. My intuition is that all models should have the attribute `self.model_parallel` because downstream logic could leverage that attribute if was universal to a transformers model. For example, we could remove the attribute from `TrainingArguments` from `Trainer` and treat the model training differently if the model was in a model parallel state when passed to the `Trainer` (and maybe we should do that now). GPT2LMHead and GPT2DoubleHead have `self.transformer`, but there isn't an analogue in other models, for example T5. What do you think, Patrick?

Reply:
Actually I think it would be a good idea of we would move this attribute the the `PretrainedModel` class and let it default to `True` and add a function `set_parallel()` to `PretrainedModel` as well. These function can then make use of the "general" set_parallel()` function. T5 would also make use of this function . What do you think @LysandreJik ?

Reply:
@patrickvonplaten @LysandreJik I think that's a great idea, Patrick. One thing to point out -- in order to make a model parallelizable, one has to go through the `forward` method of the class model and add several lines that make sure tensors are moved to the right device when `self.model_parallel` is True. It's not just the attention blocks that have to be moved to the right device. It's critical areas in the `forward` method that have to be changed for each model class when the model is parallel. If the parallelization method is added to `PretrainedModel`, we'd have to do something like return `NotImplementedError` if the user calls the parallelization method for a model which does not yet support it in its `forward` method.

Reply:
Great idea @patrickvonplaten, but maybe out of scope for this PR. Let's aim for an API we like for GPT-2, and then generalize to other models?

I'm working on a `get_layers()` utility for the gradient checkpointing (should open a PR next week), I'm sure we can re-use it for model parallelism.